### PR TITLE
#402 add mailbox state cache

### PR DESCRIPTION
**Pull request by oubchid**

### CODE REVIEW

**Review by chibenwa — COMMENTED** [review]
I read this PR but lack the underlying knowledge to approve it. Feel free to do whatever you want with it.

I perform the review because I am curious and reviews out of your technical scope should be encouraged. All works deserving an output, I left this simple comment.

Cheers.

**Comment by Arsnael** [comment]
Same as @chibenwa, it looks correct in a code point of view but I lack underlying knowledge to give a thumb up. 

Question though, I saw you make jmap calls with accountId to null... Like that for a mailbox/get:
```
mailboxesListPromise = withJmapClient(function(jmapClient) {
          return jmapClient.mailbox_get({
            accountId: null,
            ids: null
          })
```
Normally you need the accountId or you would get an error response. But maybe it's something you handle in the js jmap client lib for now? Just trying to understand here :)

Cheers

**Comment by chibenwa** [comment]
> Normally you need the accountId or you would get an error response.

The accountId is injected by the jmap-client-ts lib if missing @Arsnael

**Comment by oubchid** [comment]
@chibenwa @Arsnael 
Ok so I explain the goal here is to add `state` to cache service to be able to manage changes.

**Comment by chibenwa** [comment]
Great.

On an external point of view, this "cache" object leaks it's inside to the entire world. Maybe (ideally?) we need methods to isolates its insides to the outer world?

**Comment by alagane** [comment]
> Great.
> 
> On an external point of view, this "cache" object leaks it's inside to the entire world. Maybe (ideally?) we need methods to isolates its insides to the outer world?

I don't understand your remark. The cache is just some object stored on a running frontend on browser. It is short lived. Refreshing the page refreshes the cache.

**Comment by oubchid** [comment]
> > Great.
> > On an external point of view, this "cache" object leaks it's inside to the entire world. Maybe (ideally?) we need methods to isolates its insides to the outer world?
> 
> I don't understand your remark. The cache is just some object stored on a running frontend on browser. It is short lived. Refreshing the page refreshes the cache.

same answer

**Comment by chibenwa** [comment]
Let me rephrase: I value encapsulation.

Example of potential benefits: we decide ti change the cache structure to (say) a HashMap for whatever reasons. Or to ensure that only a set of allowed operations are performed on the array...

To be honnest, this is not a really important remark. But it might be important for differential updates...

File: src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.spec.js [hunk 1]
```diff
@@ -1491,9 +1491,12 @@ describe('The inboxMailboxesService factory', function() {
   describe('The updateSharedMailboxCache function', function() {
 
     it('should do anything on inboxMailboxesCache', function(done) {
-      inboxMailboxesCache = [];
+      inboxMailboxesCache = { state: '1', list: [] };
```
Comment (alagane) [inline src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.spec.js:1494]: Should set the cache with
```suggestion
      inboxMailboxesCache.state = '1';
      inboxMailboxesCache.list = [];
```

**Review by alagane — CHANGES_REQUESTED** [review]

File: src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js [hunk 2]
```diff
@@ -258,43 +262,55 @@ angular.module('esn.inbox.libs')
...
+              .then(function(sharedMailboxList) {
+                const sharedMailboxCache = _getSharedMailboxes(inboxMailboxesCache.list);
+                const removedSharedFoldersIds = _getDifferenceById(sharedMailboxCache, sharedMailboxList);
+
+                if (!_.isEmpty(removedSharedFoldersIds)) {
+
+                  _removeMailboxesFromCache(removedSharedFoldersIds);
+
+                  if (removedSharedFoldersIds.includes($state.params.context) === true) {
+                    $state.go('unifiedinbox.inbox', { type: '', account: '', context: '' }, { location: 'replace' });
```
Comment (chibenwa) [inline src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js:293]: How much indentation levels are acceptable?

In my knowledge, method extraction is a powerfull tool not to need 4k screens to read code.

I also allows to better name things.

My take is that maybe those callbacks could be standalone named functions to ease reading...

**Review by chibenwa — COMMENTED** [review]
No real relevant other comments to do.

File: src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js (hunk 2 above)
Comment (oubchid) [reply src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js:293]: The Indentation base used in the project is space : 2 and it depends on imbrication

Comment (chibenwa) [reply src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js:293]: No that is not what I mean. I count 10 spaces which gives 5 indentation level.

That depicts a big overall complexity of that code. I need to keep 5 branches subdivisions in mind to know what it does.

Generally we loose track from 2-3 indentation levels. 

Extracting things to named separated functions allows to bring the complexity down and enables to keep up with reading.

**Comment by chibenwa** [comment]
Force push forces to read the whole code yet-again and should be avoided.

I tend to prefer fixup commits in order not to force a full re-read for committed reviewers (`git commit --fixup aed234` were aed234 is the commit to fix)

File: src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js [hunk 3]
```diff
@@ -258,43 +262,59 @@ angular.module('esn.inbox.libs')
...
+                return _updateSharedMailboxList(sharedMailboxList);
+              })
+              .then(function() {
+                $rootScope.$broadcast(INBOX_EVENTS.SHARED_FOLDERS_UPDATED);
+
+                return _getSharedMailboxes(inboxMailboxesCache.list);
+              });
+          }
+
+          return _getSharedMailboxes(inboxMailboxesCache.list);
```
Comment (alagane) [inline src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js:294]: If those lines are put in a `then` after the changes functions, they would not have to be duplicated.

File: src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js [hunk 4]
```diff
@@ -258,43 +262,59 @@ angular.module('esn.inbox.libs')
...
+                return mailboxes.list;
+              })
+              .then(function(mailboxList) {
+                return _addSharedMailboxVisibility(_getSharedMailboxes(mailboxList));
+              })
+              .then(function(sharedMailboxList) {
+                return _updateSharedMailboxList(sharedMailboxList);
+              })
+              .then(function() {
+                $rootScope.$broadcast(INBOX_EVENTS.SHARED_FOLDERS_UPDATED);
```
Comment (alagane) [inline src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js:288]: All this part could be put in a function like `fetchMailboxesAndUpdateSharedMailboxCache`, so this would reduce complexity of the `updateSharedMailbox` function.

**Review by alagane — CHANGES_REQUESTED** [review]

File: src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js (hunk 3 above)
Comment (oubchid) [reply src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js:294]: you auto review a part of your code (pair )
what do you mean

**Comment by oubchid** [comment]
> Force push forces to read the whole code yet-again and should be avoided.
> 
> I tend to prefer fixup commits in order not to force a full re-read for committed reviewers (`git commit --fixup aed234` were aed234 is the commit to fix)
> 
> Because write now I know you changed something but I can't tell what...

Didn't know will use it

File: src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js (hunk 3 above)
Comment (oubchid) [reply src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js]: done

File: src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js (hunk 4 above)
Comment (oubchid) [reply src/esn.inbox.libs/app/services/mailboxes/mailboxes-service.js:288]: I prefer separate fetch and update rather than mix things in a function in this case 
and i'm using `_updateSharedMailboxList` so the code is more clear now

**Review by alagane — DISMISSED** [review]

**Comment by chibenwa** [comment]
Crap please stop force pushing! I can't follow your work....

**Comment by oubchid** [comment]
I'm removing a commit here to add changes on another PR so I didn't push any new changes to the last review.....
force push is helpful in this case

**Comment by chibenwa** [comment]
Pro-tip: Then leave a comment to say it so that I am not tempted to review it again ;-)

**Review by alagane — APPROVED** [review]

**Review by fabienmoyon — APPROVED** [review]
